Add missing optimization for `*`,`/` between Diagonal and Triangular

N5N3 opened the pull request:
This PR tries to add missing optimization for `*`, `/` related fuctions between `Diagonal` and `AbstractTriangular`.
It was first discussed in https://github.com/JuliaLang/julia/pull/42321#issuecomment-924581684, and I notice that such patten also applies to `/`, `ldiv!`, `\` and `rdiv!`.
I believe that unified implement help us cover all possible optimization.
For example, `lmul!` has been missing for years.

Test has been added. 

(I hope I didn't ignore anything.)

N5N3 committed:
Add missing optimization for *,/ related fuctions between Diagonal and Triangular

N5N3 committed:
Update diagonal.jl

fix typo

N5N3 committed:
Update diagonal.jl

add test

N5N3 committed:
Update diagonal.jl

replace all `AbstractArray` with `AbstractVecOrMat`

dkarrasch commented on stdlib/LinearAlgebra/src/diagonal.jl:
```diff
@@ -352,15 +316,11 @@ function mul!(C::AbstractMatrix, Da::Diagonal, Db::Diagonal, alpha::Number, beta
     return C
 end
 
+#TODO: many of /, \ related function has no size check and singular check
```
I believe they are implicit in the `[l/r]div!` or in the broadcast...?

N5N3 replied:
./ and .\ doesn't do singular check. It returns nan directly.
So at least broadcast based r/div! misses that.

For size check, broadcast is enough for 2args div.
Although we can improve the error message.
For 3 arg ldiv, the output should have the same size in the 2nd dimension with  input.
Broadcast won’t throw error for cases like 10*10 out and 10*1 in.

dkarrasch replied:
Yes, I was referring to the size checks.

LarkAnspach replied:
Checks?

LarkAnspach replied:
Why?

N5N3 replied:
If you mean size check, previous implement failed to throw with the following code:
```julia
ldiv!(randn(3,3), Diagonal(1:3), 1:3)
3×3 Matrix{Float64}:
 1.0  1.0  1.0
 1.0  1.0  1.0
 1.0  1.0  1.0
```

N5N3 replied:
Since the TODO has been removed, make this as resolved.

dkarrasch commented on stdlib/LinearAlgebra/src/diagonal.jl:
```diff
@@ -380,13 +340,59 @@ function rdiv!(A::AbstractMatrix, D::Diagonal)
+(\)(D::Diagonal, b::AbstractVector) = D.diag .\ b
+(\)(Da::Diagonal, Db::Diagonal) = Diagonal(Da.diag .\ Db.diag)
+
+ldiv!(x::AbstractVecOrMat, A::Diagonal, b::AbstractVecOrMat) = (x .= A.diag .\ b)
```
This one doesn't throw a `SingularException`. I think we have a fallback for 3-arg `ldiv!`, something like `ldiv!(copyto!(x, b), A)` or so, that admittedly has two passes over `x`, but otherwise would inherit all the exception handling from the function below. Not sure what's the right balance between code duplication and performance.

dkarrasch replied:
I just noticed that this is not a new method, just moved here... Still, I think we should decide whether we want to remove this method and let it be handled by the generic fallback
```julia
# Generic fallback. This assumes that B and Y have the same sizes.
ldiv!(Y::AbstractArray, A::AbstractMatrix, B::AbstractArray) = ldiv!(A, copyto!(Y, B))
```
or have a fully-featured 3-arg version. The fallback seems to indicate we prefer to have 2-arg `ldiv!`s.

N5N3 replied:
2-arg version is preferred in general fallback because we call LAPACK to handle `ldiv!` in most cases?
For these broadcast based methods, 3-arg version seems to be more general (and efficient)? 
So maybe we can put the kernel in 3-arg version, and call it like `ldiv!(D::Diagnal, x) = ldiv!(x,D,x) `

BTW, since we have 3-arg `ldiv!`, does it make sense to add   a 3-arg `rdiv!`?

dkarrasch replied:
I agree. Let's make the 3-arg `ldiv!` the default for `Digaonal`s and define 2-arg `ldiv!` as you say. It seems like we don't have any 3-arg `rdiv!` methods available, but 47 2-arg `rdiv!`s. So let's stick to that. The 3-arg `ldiv!` has to do with that for some factorizations one can only handle 3-arg versions, so we have a generic fallback, as well, IIRC.

N5N3 replied:
Done.
I think all `\` `/` related methods in think file do needed check, include singular and size, properly via `ldiv!` and `rdiv!`.

dkarrasch commented on stdlib/LinearAlgebra/src/diagonal.jl:
```diff
@@ -380,13 +340,59 @@ function rdiv!(A::AbstractMatrix, D::Diagonal)
+        throw(DimensionMismatch("diagonal matrix is $(length(D.diag)) by $(length(D.diag)) but right hand side has $m rows"))
+    end
+    (m == 0 || n == 0) && return B
+    for j = 1:n
```
Now this one needs one-based indexing for `B`, `D.diag` has it anyway:
```suggestion
    require_one_based_indexing(B)
    for j = 1:n
```

N5N3 replied:
added

N5N3 committed:
disambiguation

N5N3 committed:
Update diagonal.jl

N5N3 commented:
Does failure on freebsd64 related to this PR?
I see `qr` and `triangular`, but it passed once before I added the `require_one_based_indexing(B)`.
It seems win32 and win64's tests are not running?

dkarrasch commented:
freebsd64 can be ignored, if this is the only one failing.

N5N3 committed:
add more sigular check

all `\` and `/` like calls fall back to method with this check now.

N5N3 committed:
whitespace

dkarrasch commented on stdlib/LinearAlgebra/src/diagonal.jl:
```diff
@@ -566,11 +539,17 @@ function svd(D::Diagonal{<:Number})
+*(x::TransposeAbsVec, D::Diagonal) = Transpose(map((t,s) -> transpose(t)*s, D.diag, parent(x)))
+*(x::AdjointAbsVec,   D::Diagonal, y::AbstractVector) = _mapreduce_prod(*, x, D, y)
+*(x::TransposeAbsVec, D::Diagonal, y::AbstractVector) = _mapreduce_prod(*, x, D, y)
+/(u::AdjointAbsVec, D::Diagonal) = adjoint(adjoint(D) \ u.parent)
+/(u::TransposeAbsVec, D::Diagonal) = transpose(transpose(D) \ u.parent)
```
How come we introduced new ambiguous methods here?

N5N3 replied:
`(/)(A::AbstractVecOrMat, D::Diagonal)` was `(/)(A::StrideMatrix, D::Diagonal)`

dkarrasch commented on stdlib/LinearAlgebra/src/diagonal.jl:
```diff
@@ -380,13 +340,50 @@ function rdiv!(A::AbstractMatrix, D::Diagonal)
+    m′, n′ = size(B, 1), size(B, 2)
+    m == d || throw(DimensionMismatch("right hand side has $m rows but D is $d by $d"))
+    (m, n) == (m′, n′) || throw(DimensionMismatch("expect output to be $m by $n, but got $m′ by $m′"))
+    0 in D.diag && throw(SingularException(findfirst(iszero, D.diag)))
```
How about making this a single-pass operation? Something like
```julia
j = findfirst(iszero, D.diag)
j != nothing && throw(SingularException(j))
```
?

dkarrasch replied:
Or write out the loops as we used to have them...

N5N3 replied:
I do this because `0 in abstractrange` has an optimized fallback.
I think this does no harm to runtime performace, as for diag without zero, `findfirst` won't be called.

dkarrasch reviewed (commented):
A few more comments. Otherwise, this is shaping up nicely.

N5N3 commented on stdlib/LinearAlgebra/src/diagonal.jl:
```diff
@@ -352,19 +317,14 @@ function mul!(C::AbstractMatrix, Da::Diagonal, Db::Diagonal, alpha::Number, beta
-end
+(/)(A::AbstractVecOrMat, D::Diagonal) =
+    rdiv!(copy_similar(A, promote_op(/, eltype(A), eltype(D))), D)
+(/)(Da::Diagonal, Db::Diagonal) = Diagonal(Db \ Da.diag)
```
This and L345 was changed to make it call `ldiv!` rather than broadcast directly to reuse the input checks.

I just realize that the following is correct only for `Eltype <: Number`
```julia
A / B = A * B⁻¹ = (B⁻¹ᵀ * Aᵀ)ᵀ = B⁻¹ᵀ * Aᵀ = B⁻¹ * A = B \ A
```
since `/` perform copy while `\` not, shall we replace this with
```julia
(/)(Da::Diagonal{<:Number}, Db::Diagonal{<:Number}) = Diagonal(Db \ Da.diag)
(/)(Da::Diagonal, Db::Diagonal) = Diagonal(Da.diag ./ Db.diag)
```

dkarrasch replied:
Excellent point about the `SingularException` check! But isn't this just `Diagonal(Da.diag / Db)`? I'm not sure the code bloat (with `_rdiv!`) is worth it here. Line 345 is able to avoid the first copy-pass "for free" (due to the 3-arg `ldiv!`).

N5N3 replied:
In fact I tried that, and then notice that `Da.diag/Db` only works for 1*1 Db.

dkarrasch replied:
🤦 True!

dkarrasch replied:
> I just realize that the following is correct only for `Eltype <: Number`

This is not even correct for that case! The third step is wrong, you can't just omit the outer transpose.

N5N3 replied:
That should be ok?
Since `(B⁻¹ᵀ * Aᵀ)` isa `Diagnal` and we have `transpose(D::Diagonal{<:Number}) = D`

N5N3 commented:
Error found in
```julia
@test Diagonal(spzeros(5)) \ view(rand(10), 1:5) == [Inf, Inf, Inf, Inf, Inf]
```
This throw `SingularException(1)` after this PR.
Just change it to `@test_throws` ?

Others are unexpected pass
```julia
@test H/x == Array(H)/x broken = eltype(H) <: Furlong && x isa Union{Bidiagonal, Diagonal, UpperTriangular}
@test x\H == x\Array(H) broken = eltype(H) <: Furlong && x isa Union{Bidiagonal, Diagonal, UpperTriangular}
@test H/x isa UpperHessenberg broken = eltype(H) <: Furlong && x isa Union{Bidiagonal, Diagonal}
@test x\H isa UpperHessenberg broken = eltype(H) <: Furlong && x isa Union{Bidiagonal, Diagonal}
```
All these have `Diagonal` but i haven't look into the reason.
Is broken the excepted behaviour? Or we just remove `Diagonal` here.

dkarrasch commented:
> This throw `SingularException(1)` after this PR.
> Just change it to `@test_throws` ?

I'd say yes, but one could argue that diagonal solves are nothing but a sequence of scalar divisions. When we have scalar `5/0` we get `Inf` without a `SingularException` or any other warning. In any case, we need to make sure that
```julia
@test Diagonal(spzeros(5)) \ view(rand(10), 1:5) == Matrix(Diagonal(spzeros(5))) \ view(rand(10), 1:5)
```
If we change this for diagonal, this is going to be passed on the the matrix one, because it checks for triangularity:
```julia
function (\)(A::AbstractMatrix, B::AbstractVecOrMat)
    require_one_based_indexing(A, B)
    m, n = size(A)
    if m == n
        if istril(A)
            if istriu(A)
                return Diagonal(A) \ B
...
```
I'd vote for consistency with
```julia
julia> lu(zeros(5,5))
ERROR: SingularException(1)
```
so with what you have.

As for the second issue, I guess you resolved broken cases along the way. Congrats! 🎉  🤣

N5N3 committed:
Update diagonal.jl

1. fix error message
2. fix Diagnal / Diagnal
3. omit `0 in` check
4. add a 3-arg `_rdiv!` to avoid copy.

N5N3 committed:
Update hessenberg.jl

N5N3 committed:
Update higherorderfns.jl

N5N3 commented on stdlib/LinearAlgebra/src/diagonal.jl:
```diff
@@ -380,13 +340,50 @@ function rdiv!(A::AbstractMatrix, D::Diagonal)
+    m, n = size(A, 1), size(A, 2)
+    m′, n′ = size(B, 1), size(B, 2)
+    m == d || throw(DimensionMismatch("right hand side has $m rows but D is $d by $d"))
+    (m, n) == (m′, n′) || throw(DimensionMismatch("expect output to be $m by $n, but got $m′ by $m′"))
```
fixed

dkarrasch commented on stdlib/LinearAlgebra/src/diagonal.jl:
```diff
@@ -352,19 +317,16 @@ function mul!(C::AbstractMatrix, Da::Diagonal, Db::Diagonal, alpha::Number, beta
-ldiv!(x::AbstractArray, A::Diagonal, b::AbstractArray) = (x .= A.diag .\ b)
+(/)(A::AbstractVecOrMat, D::Diagonal) =
+    _rdiv!(similar(A, promote_op(/, eltype(A), eltype(D)), size(A)), A, D)
+(/)(Da::Diagonal, Db::Diagonal) = Diagonal(vec(permutedims(Da.diag) / Db))
```
What's the performance of this both for small and large cases? Once we're at it, maybe we have another `_rdiv!` method for the two-Diagonal case? It should be quite a couple of lines shorter than the general one.

N5N3 replied:
At least `permutedims` should do harm to small case.
I also wonder whether the compile know the first dimension is one sized. If not there should be more performance regression.

Implement this in the deepest fallback call is a good idea. As `rdiv!` for this case should also be included automatically. Similar for 3arg `ldiv!`

N5N3 replied:
Found that `similar(::Diagnal, ::Type, size(A))` returns a sparse matrix.
Is there any reason we use this rather than `similar(::Diagnal, ::Type)`?

dkarrasch reviewed (commented):
Since this PR almost preserves the number of lines of code, I think it is acceptable to go the last bit and avoid the `vec(permutedims` construct by some straightforward loop based code. I especially like the fact that this removes a couple of broken tests. Really awesome!

dkarrasch commented:
Up to one last suggestion, this is ready for review. The only potentially controversial aspect that I see is the change from "blind" division to a checked solve with throwing `SingularException`s.

jishnub commented on stdlib/LinearAlgebra/src/diagonal.jl:
```diff
@@ -214,11 +214,6 @@ function (*)(D::Diagonal, V::AbstractVector)
     return D.diag .* V
 end
 
-(*)(A::AbstractTriangular, D::Diagonal) =
```
Wouldn't  this still be needed for custom subtypes of `AbstractTriangular` other than the ones defined in `LinearAlgebra`?

N5N3 replied:
We have them in L552, which invoke to the unoptimized version .

N5N3 committed:
Update diagonal.jl

N5N3 committed:
Update diagonal.jl

N5N3 committed:
Update diagonal.jl

N5N3 committed:
Update diagonal.jl

add type check

N5N3 committed:
Update hessenberg.jl

N5N3 committed:
Update diagonal.jl

N5N3 commented:
The failure in `tester_win32` shows that:
```julia
      From worker 8:	
      From worker 8:	Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
      From worker 8:	Exception: EXCEPTION_ACCESS_VIOLATION at 0x156d8264 -- .text at C:\buildbot\worker-tabularasa\tester_win32\build\bin\libopenblas.dll (unknown line)
      From worker 8:	in expression starting at C:\buildbot\worker-tabularasa\tester_win32\build\share\julia\stdlib\v1.8\LinearAlgebra\test\tridiag.jl:28
      From worker 8:	.text at C:\buildbot\worker-tabularasa\tester_win32\build\bin\libopenblas.dll (unknown line)
      From worker 8:	.text at C:\buildbot\worker-tabularasa\tester_win32\build\bin\libopenblas.dll (unknown line)
      From worker 8:	.text at C:\buildbot\worker-tabularasa\tester_win32\build\bin\libopenblas.dll (unknown line)
      From worker 8:	stegr! at C:\buildbot\worker\package_win32\build\usr\share\julia\stdlib\v1.8\LinearAlgebra\src\lapack.jl:3853
      From worker 8:	eigen! at C:\buildbot\worker\package_win32\build\usr\share\julia\stdlib\v1.8\LinearAlgebra\src\tridiag.jl:294 [inlined]
      From worker 8:	eigen at C:\buildbot\worker\package_win32\build\usr\share\julia\stdlib\v1.8\LinearAlgebra\src\tridiag.jl:296
      From worker 8:	unknown function (ip: 016c7eed)
```
I found `tridiag.jl:294` is:
https://github.com/JuliaLang/julia/blob/e92545043d5c541c8f2951719be5647be0cc9ea2/stdlib/LinearAlgebra/test/tridiag.jl#L294

Can understand what happened. Seems to be a openblas issue?

dkarrasch reviewed (approved):
I'm not sure I like the `_promote_dotop` thing, but otherwise this LGTM. ;-)

N5N3 committed:
Update diagonal.jl

1. add optimized 5-arg `mul!` fallbacks, and make 3-arg `mul!` call them.
2. drop `_promote_dotop`

N5N3 committed:
Update diagonal.jl

add 5-arg `mul!` test

N5N3 commented:
I add 5-arg `mul!` optimization between Triangularand Diagonal.
The problem is how to handle `UnitUpper/Lower` input.
Currently the output's diagonal is first cached in to a `diag′::Vector` if `β != 0`.
And the output's diagonal is updated with 
```julia
for i in 1:length(x)
    data[i,i] = MulAddMul(α, β)(D.diag[i], diag′)
end
```
in the final stage.
The performance influence is small at least for `N >= 100`.
Some benchmarks:
```julia
n = 100
UTA = UnitUpperTriangular(zeros(n,n));
TA = UpperTriangular(zeros(n,n));
D = Diagonal(zeros(n));
B = TA * D;
@btime mul!($B,$D,$UTA);    # 1.640 μs (0 allocations: 0 bytes)
@btime mul!($B,$D,$TA);       #   1.530 μs (0 allocations: 0 bytes)
@btime mul!($B,$D,$UTA, $0, $1);   #   5.867 μs (5 allocations: 1008 bytes)
@btime mul!($B,$D,$TA, $0, $1);    #   5.567 μs (3 allocations: 80 bytes)
@btime mul!($B,$D,$UTA, $2, $1);   #   7.250 μs (6 allocations: 1.03 KiB)
@btime mul!($B,$D,$TA, $2, $1);   #   6.900 μs (4 allocations: 128 bytes)
```

N5N3 committed:
Update diagonal.jl

whitespace

N5N3 commented on stdlib/LinearAlgebra/src/diagonal.jl:
```diff
@@ -382,19 +380,32 @@ for Tri in (:UpperTriangular, :LowerTriangular)
+    @eval mul!(out::$Tri, A::Union{$Tri,$UTri}, D::Diagonal) = mul!(out, A, D, true, false)
+    # 5-arg mul!
+    @eval @inline mul!(out::$Tri, D::Diagonal, A::$Tri, α::Number, β::Number) =
+        $Tri(mul!(out.data, D, A.data, α, β))
```
Since 
```julia
julia> let a = randn(10,10)
           UpperTriangular(a) === UpperTriangular(a)
       end
true
```
I think it's OK? And `rmul!` was implemented as:
```julia
rmul!(A::Union{LowerTriangular,UpperTriangular}, D::Diagonal) = typeof(A)(rmul!(A.data, D))
```
I'll add these tests anyway.

N5N3 committed:
Update diagonal.jl

add === test

N5N3 committed:
Update diagonal.jl

make mul! respect alpha == 0

N5N3 committed:
white space

dkarrasch commented:
@andreasnoack Do you think there's anything wrong about this change in behavior? https://github.com/JuliaLang/julia/pull/42343#issuecomment-926769992

Absent objections, I'd merge this soon.
